[UnifiedPDF] [iOS] Implement page number indicator

aprotyas opened the pull request:
#### f1e6113a523e164159030fd1eb8dd96f33850729
<pre>
[UnifiedPDF] [iOS] Implement page number indicator
<a href="https://bugs.webkit.org/show_bug.cgi?id=288487">https://bugs.webkit.org/show_bug.cgi?id=288487</a>
<a href="https://rdar.apple.com/118154868">rdar://118154868</a>

Reviewed by Wenson Hsieh and Tim Horton.

This patch implements the page number indicator for main frame PDF
plugins on iOS. More details in-line.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

The web preference may be helpful to set (or unset) when authoring tests.
Currently, we enable it by default on iOS and gate creation of the
indicator behind it.

* Source/WTF/wtf/PlatformEnableCocoa.h:

Introduce a new compile-time guard for the page number indicator,
limiting behavior changes to the iOS family.

* Source/WebCore/en.lproj/Localizable.strings:

Localize the label text for the indicator.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::frameViewLayoutOrVisualViewportChanged):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::layoutOrVisualViewportChanged):

Set up plumbing to forward (updated) unobscured content rect to any
relevant main frame plugin, downstream of visible content rect updates.

* Source/WebKit/PlatformMac.cmake:

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _createPDFPageNumberIndicator:withFrame:pageCount:]):
(-[WKWebView _removePDFPageNumberIndicator:]):
(-[WKWebView _updatePDFPageNumberIndicator:withFrame:]):
(-[WKWebView _updatePDFPageNumberIndicator:currentPage:]):
(-[WKWebView _removeAnyPDFPageNumberIndicator]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Maintain a page number indicator (associated to a certain plugin) and
modify said indicator as necessary.

* Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.h: Added.
* Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm: Added.
(-[WKPDFPageNumberIndicator initWithFrame:view:pageCount:]):
(-[WKPDFPageNumberIndicator dealloc]):
(-[WKPDFPageNumberIndicator updatePosition:]):
(-[WKPDFPageNumberIndicator setCurrentPageNumber:]):
(-[WKPDFPageNumberIndicator setPageCount:]):
(-[WKPDFPageNumberIndicator show]):
(-[WKPDFPageNumberIndicator hide:]):
(-[WKPDFPageNumberIndicator _moveToPoint:animated:completionHandler:]):
(-[WKPDFPageNumberIndicator sizeThatFits:]):
(-[WKPDFPageNumberIndicator _updateLabel:]):

The implementation for the page number indicator. Things of note
include:

- We use a UIVisualEffectView to apply a blur style on the indicator.
- We maintain a timer that we use to present/dismiss the indicator
  predictably.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::resetStateAfterProcessTermination):

Destroy any lingering page number indicator instances as necessary.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::createPDFPageNumberIndicator):
(WebKit::PageClientImpl::removePDFPageNumberIndicator):
(WebKit::PageClientImpl::updatePDFPageNumberIndicatorLocation):
(WebKit::PageClientImpl::updatePDFPageNumberIndicatorCurrentPage):
(WebKit::PageClientImpl::removeAnyPDFPageNumberIndicator):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::createPDFPageNumberIndicator):
(WebKit::WebPageProxy::removePDFPageNumberIndicator):
(WebKit::WebPageProxy::updatePDFPageNumberIndicatorLocation):
(WebKit::WebPageProxy::updatePDFPageNumberIndicatorCurrentPage):

Plumbing information flow from the plugin to the UI process to
(eventually) WKPDFPageNumberIndicator.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::teardown):
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::attemptToUnlockPDF):
(WebKit::UnifiedPDFPlugin::geometryDidChange):
(WebKit::UnifiedPDFPlugin::visibilityDidChange):
(WebKit::UnifiedPDFPlugin::frameForPageNumberIndicatorInRootViewCoordinates const):
(WebKit::UnifiedPDFPlugin::pageNumberIndicatorEnabled const):
(WebKit::UnifiedPDFPlugin::shouldShowPageNumberIndicator const):

There is a laundry list of conditions under which we do not want to show
the page number indicator, which includes:

- If the page number indicator preference is disabled.
- When the document is locked.
- For non-full-main-frame plugins.
- If the PluginView is not visible.
- If the PDF document is not installed yet.

(WebKit::UnifiedPDFPlugin::updatePageNumberIndicatorVisibility):
(WebKit::UnifiedPDFPlugin::updatePageNumberIndicatorLocation):
(WebKit::UnifiedPDFPlugin::updatePageNumberIndicatorCurrentPage):

The current page number can be deduced if we have the unobscured content
rect. We either get this plumbed to us from local frame view viewport
change updates, or we read this actively off of the local frame view as
neessary.

(WebKit::UnifiedPDFPlugin::updatePageNumberIndicator):
(WebKit::UnifiedPDFPlugin::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::createPDFPageNumberIndicator):
(WebKit::WebPage::updatePDFPageNumberIndicatorLocation):
(WebKit::WebPage::updatePDFPageNumberIndicatorCurrentPage):
(WebKit::WebPage::removePDFPageNumberIndicator):

More plumbing.

Canonical link: <a href="https://commits.webkit.org/291204@main">https://commits.webkit.org/291204@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52e2a8b0ce72fadde69a0fc79c3ae7de58422bdb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92208 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1389 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42751 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94258 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20221 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28181 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95209 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9173 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8892 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1196 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84953 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79213 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99243 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90904 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19292 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/99243 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19544 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79443 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99243 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19581 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23524 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12296 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19273 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24445 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113527 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18965 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32845 "") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22423 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20708 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm:
```diff
@@ -0,0 +1,207 @@
+    _webView = view;
+
+    [self setTranslatesAutoresizingMaskIntoConstraints:NO];
+    self.translatesAutoresizingMaskIntoConstraints = NO;
```
Is this necessary? I don't see any use of auto layout in this patch?

aprotyas replied:
not necessary, I will remove it

smfr commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -88,6 +88,9 @@
 "%d files" = "%d files";
 
+/* Label for PDF page number indicator. */
+"%d of %d" = "%d of %d";
```
This probably won't  be localizable in some languages.

aprotyas replied:
That's possible, yes, I'm not fully sure how to evaluate this. I just lifted our existing label text.

aprotyas replied:
Fixed by adopting positional arguments!

smfr commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -429,6 +434,10 @@ struct PerWebProcessState {
     RetainPtr<WKScrollGeometry> _currentScrollGeometry;
+
+#if ENABLE(PDF_PAGE_NUMBER_INDICATOR)
+    WTF::HashMap<WebKit::PDFPluginIdentifier, RetainPtr<WKPDFPageNumberIndicator>> _pdfPageNumberIndicators;
```
Why would there be more than one? If may one OK to parent the main frame one directly under the web view, but you can't do that for subframes because the indicator won't scroll with the frame.

aprotyas replied:
Indeed. I originally wrote it this way hoping it would simplify the embedded PDF case, but it won't.

I've changed this to just be a `std::pair<Markable<PluginIdentifier>, RetainPtr<WKPDFPageNumberIndicator>>` instead.

pxlcoder commented on Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm:
```diff
@@ -0,0 +1,207 @@
+    [_label setAlpha:indicatorLabelOpacity];
+    [_label setAdjustsFontSizeToFitWidth:YES];
+    [_label setAutoresizingMask:UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight];
+    [_label setTranslatesAutoresizingMaskIntoConstraints:YES];
```
(not necessary, as the default is true)

aprotyas replied:
Removed

pxlcoder commented on Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm:
```diff
@@ -0,0 +1,207 @@
+    [_label setBackgroundColor:nil];
+    [_label setTextAlignment:NSTextAlignmentCenter];
+    [_label setFont:[UIFont boldSystemFontOfSize:indicatorFontSize]];
+    [_label setTextColor:[UIColor blackColor]];
```
Is it supposed to be black in both light and dark mode?

pxlcoder replied:
(chatted about this in person, existing implementation is "always light mode", so this should be too)

aprotyas commented on Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm:
```diff
@@ -0,0 +1,201 @@
+
+    auto animations = [view = retainPtr(self), point, completionHandler = WTFMove(completionHandler)] mutable {
+        CGRect frame = [view frame];
+        bool originMoved = static_cast<bool>(CGPointEqualToPoint(frame.origin, point) == NO);
```
This `static_cast` is unnecessary. I'll remove it

whsieh replied:
Yeah, I think this should be sufficient:

```suggestion
        bool originMoved = !CGPointEqualToPoint(frame.origin, point);
```

aprotyas replied:
Fixed!

aprotyas commented on Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm:
```diff
@@ -0,0 +1,201 @@
+    }];
+
+    if (_timer)
+        [_timer setFireDate:[NSDate dateWithTimeIntervalSinceNow:indicatorTimeout.seconds()]];
```
I wish `WTF::Seconds::operator NSTimeInterval() const { return seconds(); }` was a thing 😞 I'll add it in a future commit

whsieh commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3250,6 +3250,9 @@ void LocalFrameView::layoutOrVisualViewportChanged()
             scrollingCoordinator->frameViewVisualViewportChanged(*this);
     }
+
+    if (Page* page = m_frame->page())
```
```suggestion
    if (RefPtr page = m_frame->page())
```

aprotyas replied:
Fixed!

whsieh commented on Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm:
```diff
@@ -0,0 +1,201 @@
+    };
+
+    if (animated)
+        [UIView animateWithDuration:indicatorMoveDuration.seconds() animations:makeBlockPtr(WTFMove(animations)).get()];
```
Just to double check, we expect this completion handler to be called when the animation begins (not when it ends)?

For the latter, I think we'd need to call `+animateWithDuration:animations:completion:`

aprotyas replied:
I did indeed want the handler to run at the end of animations, not at the beginning. Good catch! I used `+animateWithDuration:animations:completion:` instead and changed up the completion handler signature a little to still make use of the `originMoved` information.

aprotyas committed:
[UnifiedPDF] [iOS] Implement page number indicator
https://bugs.webkit.org/show_bug.cgi?id=288487
rdar://118154868

Reviewed by Wenson Hsieh and Tim Horton.

This patch implements the page number indicator for main frame PDF
plugins on iOS. More details in-line.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

The web preference may be helpful to set (or unset) when authoring tests.
Currently, we enable it by default on iOS and gate creation of the
indicator behind it.

* Source/WTF/wtf/PlatformEnableCocoa.h:

Introduce a new compile-time guard for the page number indicator,
limiting behavior changes to the iOS family.

* Source/WebCore/en.lproj/Localizable.strings:

Localize the label text for the indicator.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::frameViewLayoutOrVisualViewportChanged):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::layoutOrVisualViewportChanged):

Set up plumbing to forward (updated) unobscured content rect to any
relevant main frame plugin, downstream of visible content rect updates.

* Source/WebKit/PlatformMac.cmake:

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _createPDFPageNumberIndicator:withFrame:pageCount:]):
(-[WKWebView _removePDFPageNumberIndicator:]):
(-[WKWebView _updatePDFPageNumberIndicator:withFrame:]):
(-[WKWebView _updatePDFPageNumberIndicator:currentPage:]):
(-[WKWebView _removeAnyPDFPageNumberIndicator]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Maintain a page number indicator (associated to a certain plugin) and
modify said indicator as necessary.

* Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.h: Added.
* Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm: Added.
(-[WKPDFPageNumberIndicator initWithFrame:view:pageCount:]):
(-[WKPDFPageNumberIndicator dealloc]):
(-[WKPDFPageNumberIndicator updatePosition:]):
(-[WKPDFPageNumberIndicator setCurrentPageNumber:]):
(-[WKPDFPageNumberIndicator setPageCount:]):
(-[WKPDFPageNumberIndicator show]):
(-[WKPDFPageNumberIndicator hide:]):
(-[WKPDFPageNumberIndicator _moveToPoint:animated:completionHandler:]):
(-[WKPDFPageNumberIndicator sizeThatFits:]):
(-[WKPDFPageNumberIndicator _updateLabel:]):

The implementation for the page number indicator. Things of note
include:

- We use a UIVisualEffectView to apply a blur style on the indicator.
- We maintain a timer that we use to present/dismiss the indicator
  predictably.

* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::resetStateAfterProcessTermination):

Destroy any lingering page number indicator instances as necessary.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::createPDFPageNumberIndicator):
(WebKit::PageClientImpl::removePDFPageNumberIndicator):
(WebKit::PageClientImpl::updatePDFPageNumberIndicatorLocation):
(WebKit::PageClientImpl::updatePDFPageNumberIndicatorCurrentPage):
(WebKit::PageClientImpl::removeAnyPDFPageNumberIndicator):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::createPDFPageNumberIndicator):
(WebKit::WebPageProxy::removePDFPageNumberIndicator):
(WebKit::WebPageProxy::updatePDFPageNumberIndicatorLocation):
(WebKit::WebPageProxy::updatePDFPageNumberIndicatorCurrentPage):

Plumbing information flow from the plugin to the UI process to
(eventually) WKPDFPageNumberIndicator.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::teardown):
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::attemptToUnlockPDF):
(WebKit::UnifiedPDFPlugin::geometryDidChange):
(WebKit::UnifiedPDFPlugin::visibilityDidChange):
(WebKit::UnifiedPDFPlugin::frameForPageNumberIndicatorInRootViewCoordinates const):
(WebKit::UnifiedPDFPlugin::pageNumberIndicatorEnabled const):
(WebKit::UnifiedPDFPlugin::shouldShowPageNumberIndicator const):

There is a laundry list of conditions under which we do not want to show
the page number indicator, which includes:

- If the page number indicator preference is disabled.
- When the document is locked.
- For non-full-main-frame plugins.
- If the PluginView is not visible.
- If the PDF document is not installed yet.

(WebKit::UnifiedPDFPlugin::updatePageNumberIndicatorVisibility):
(WebKit::UnifiedPDFPlugin::updatePageNumberIndicatorLocation):
(WebKit::UnifiedPDFPlugin::updatePageNumberIndicatorCurrentPage):

The current page number can be deduced if we have the unobscured content
rect. We either get this plumbed to us from local frame view viewport
change updates, or we read this actively off of the local frame view as
neessary.

(WebKit::UnifiedPDFPlugin::updatePageNumberIndicator):
(WebKit::UnifiedPDFPlugin::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::frameViewLayoutOrVisualViewportChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::createPDFPageNumberIndicator):
(WebKit::WebPage::updatePDFPageNumberIndicatorLocation):
(WebKit::WebPage::updatePDFPageNumberIndicatorCurrentPage):
(WebKit::WebPage::removePDFPageNumberIndicator):

More plumbing.

Canonical link: https://commits.webkit.org/291204@main
